app/models/ontology/sentences.rb

```diff
@@ -1,135 +1,151 @@
+              sentence_ids.include?(sentence.id)
+            end
+            other_translated.each { |ot| arr.first << ot }
+            other_sentences.each { |os| arr.last << os }
+            arr
+          end
+        [translated + imported.first, self.sentences + imported.last]
       end
-      [translated + imported.first, self.sentences + imported.last]
-    end
 
-    # Find import-links which describe the following mapping:
-    # some ontology imports self.
-    def incoming_imports_with_mappings
-      # INNER JOINS always return an empty result set if the
-      # ON clause does not match.
-      Link.joins(:entity_mappings).where(source_id: self, kind: 'import')
-    end
+      # Find import-mappings which describe the following mapping:
+      # some ontology imports self.
+      def incoming_imports_with_mappings
+        # INNER JOINS always return an empty result set if the
+        # ON clause does not match.
+        Mapping.joins(:symbol_mappings).where(source_id: self, kind: 'import')
+      end
 
-    def create_translated_sentences
-      query, args = links_by_kind_query(self, 'import')
-      imported_ontology_ids = pluck_select([query, *args], :ontology_id).reverse
-      imported_ontology_ids.each { |o_id| Ontology.find(o_id).create_translated_sentences }
-      import_links = incoming_imports_with_mappings
-      if import_links.any?
-        combined_sentences.each do |sentence|
-          import_links.each do |import_link|
-            import_link.entity_mappings.each do |mapping|
-              translate_sentence(sentence, mapping)
+      def create_translated_sentences
+        query, args = mappings_by_kind_query(self, 'import')
+        imported_ontology_ids =
+          pluck_select([query, *args], :ontology_id).reverse
+        imported_ontology_ids.each do |o_id|
+          Ontology.find(o_id).create_translated_sentences
+        end
+        import_mappings = incoming_imports_with_mappings
+        if import_mappings.any?
+          combined_sentences.each do |sentence|
+            import_mappings.each do |import_mapping|
+              import_mapping.symbol_mappings.each do |mapping|
+                translate_sentence(sentence, mapping)
+              end
             end
           end
         end
       end
-    end
 
-    def translate_sentence(sentence, mapping)
-      applicable_sentence = TranslatedSentence.choose_applicable(sentence, mapping)
-      if mapping.applicable?(applicable_sentence)
-        translated_text = mapping.apply(applicable_sentence)
-        audience = mapping.link.target
-        translated_sentence = TranslatedSentence.where(
-          entity_mapping_id: mapping,
-          audience_id: audience,
-          sentence_id: sentence,
-          ontology_id: sentence.ontology).first_or_initialize
-        translated_sentence.translated_text = translated_text
-        translated_sentence.save
-        translated_sentence
+      def translate_sentence(sentence, mapping)
```

Comment:
Well the TranslatedSentence creation could be extracted into it's own method, but this shouldn't necessarily be done in this branch. Maybe create an issue for refactoring large methods of TranslatedSentence.